Support `Get-PSHostProcessInfo` and `Enter-PSHostProcess` on non-Windows

SteveL-MSFT opened the pull request:
## PR Summary

There a some differences in support of named pipes for Windows and non-Windows.  Named pipes on Unix have a 104 character path limit.  On macOS, the `$env:TMPDIR` (on my system) is already 49 characters; corefx adds 12 more.  Since AppDomainName isn't really used, changed it from `DefaultAppDomain` to `None` to shorten the name.  Need to keep it since Windows PowerShell expects it.  Changed `starttime` part of pipe name to 8 hex characters which is to provide uniqueness to the pipe name.

Fix https://github.com/PowerShell/PowerShell/issues/5110
Possibly also fix 6435, but need verification. https://github.com/PowerShell/PowerShell/issues/6435

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [ ] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [ ] Not Applicable or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

SteveL-MSFT committed:
initial code to add UNIX path to named pipes

SteveL-MSFT committed:
fix codefactor and test failure

Commit:
remove unnecessary legacy Windows specific doe

SteveL-MSFT committed:
[feature]

cleanup existing Windows specific code and update tests

iSazonov commented on src/System.Management.Automation/engine/remoting/commands/EnterPSHostProcessCommand.cs:
```diff
@@ -535,9 +544,13 @@ internal static IReadOnlyCollection<PSHostProcessInfo> GetAppDomainNamesFromProc
                 {
                     // This is a PowerShell named pipe.  Parse the process Id, AppDomain name, and process name.
+#if !UNIX
                     int pStartTimeIndex = namedPipe.IndexOf(".", startIndex, StringComparison.OrdinalIgnoreCase);
```
We should replace `"."` string with char `'.'
```suggestion
                    int pStartTimeIndex = namedPipe.IndexOf('.', startIndex, StringComparison.OrdinalIgnoreCase);
```

iSazonov commented on src/System.Management.Automation/engine/remoting/commands/EnterPSHostProcessCommand.cs:
```diff
@@ -535,9 +544,13 @@ internal static IReadOnlyCollection<PSHostProcessInfo> GetAppDomainNamesFromProc
+#else
+                        int pStartTimeIndex = 0;
+#endif
                         int pIdIndex = namedPipe.IndexOf(".", pStartTimeIndex + 1, StringComparison.OrdinalIgnoreCase);
```
```suggestion
                        int pIdIndex = namedPipe.IndexOf('.', pStartTimeIndex + 1, StringComparison.OrdinalIgnoreCase);
```

iSazonov commented on src/System.Management.Automation/engine/remoting/common/RemoteSessionNamedPipe.cs:
```diff
@@ -590,6 +611,11 @@ public void Dispose()
 
         internal static CommonSecurityDescriptor GetServerPipeSecurity()
         {
+            if (!Platform.IsWindows)
```
Why runtime check? We could use `#if UNIX`.

SteveL-MSFT replied:
I think originally I left this to see if this method made sense on UNIX, but it's for allowing admins to enter a user's runspace.  For UNIX, I think we should just only allow same user.  I'll update the code to `#if`.

SteveL-MSFT committed:
address Ilya's feedback, some minor cleanup

SteveL-MSFT committed:
[feature]
use same model as AMSI uninitialize to cleanup named pipes on exit
when enumerating pshostprocessinfo, check if they are valid, otherwise exclude and try to remove

PaulHigin commented:
@SteveL-MSFT 

Regarding the server named pipe dispose, this is done in WindowsPowerShell via the CLR AppDomain unload event, which is not supported on CORECLR.  However, you can use the AppDomain.CurrentDomain.ProcessExit event that we use for AMSI un-initializing.  Unfortunately, these CLR events are not reliable, but they are better than nothing.

The start time added to the filename is intended to make the named pipe instance unique, to prevent a DOS or man-in-the-middle attack.  We will want to keep something like this, even if it means using something shorter.

SteveL-MSFT committed:
[feature]
address Paul's feedback

TylerLeonhardt reviewed (approved):
LGTM 👍

TylerLeonhardt commented:
I love when a PR has more deletions than insertions but more functionality

SteveL-MSFT commented on src/System.Management.Automation/engine/remoting/commands/EnterPSHostProcessCommand.cs:
```diff
@@ -565,14 +573,31 @@ internal static IReadOnlyCollection<PSHostProcessInfo> GetAppDomainNamesFromProc
-                                        new PSHostProcessInfo(pName, id, appDomainName));
+                                    // If SMA.dll did not exit cleanly, there could be left over pipes
+                                    // so we check if the process exists and matches the name, otherwise remove.
+                                    var process = System.Diagnostics.Process.GetProcessById(id);
```
Unfortunately, `Process` is a parameter therefore a property so some of the uses have to be fully qualified like this one

SteveL-MSFT committed:
[feature]
address Paul and Tyler's feedback

SteveL-MSFT committed:
[feature]
added tests

PaulHigin commented on src/System.Management.Automation/engine/hostifaces/LocalConnection.cs:
```diff
@@ -836,6 +836,12 @@ private void DoCloseHelper()
                     }
                 }
 
+                if (s_IPCNamedPipeServer != null)
```
I don't think we want to dispose here ... hosted scenarios can have no runspaces but we still want to be able to connect to them.  What I had in mind was doing this on the `AppDomain.CurrentDomain.ProcessExit += CurrentDomain_ProcessExit;` event, but I don't know if this is supported on non-Windows platforms.  But if there is an event like this where the CLR calls it on exit, that would be best.

Another option is to clean up old pipe files on start up, in a separate thread from the main thread.

PaulHigin commented on src/System.Management.Automation/engine/remoting/common/RemoteSessionNamedPipe.cs:
```diff
@@ -26,7 +27,13 @@ internal static class NamedPipeUtils
         internal const string DefaultAppDomainName = "DefaultAppDomain";
         internal const string NamedPipeNamePrefix = "PSHost.";
+#if UNIX
+        internal const string NamedPipeNamePrefixSearch = "CoreFxPipe_PSHost*";
```
If length is an issue, we could make this prefix smaller, and remove the domain name.

PaulHigin reviewed (changes requested):
Overall, this is great and I am really glad this feature is getting ported to Linux/MacOS platforms.  If you want to add tests, you can use the Runspace API to test the feature:
```powershell
# Create connection info using target process Id
$npInfo = [System.Management.Automation.Runspaces.NamedPipeConnectionInfo]::new(14376)
$rs = [runspacefactory]::CreateRunspace($npInfo)
$rs.Open()
$ps = [powershell]::Create()
$ps.Runspace = $rs
$ps.AddScript('"PID = $pid"').Invoke()
$rs.Dispose()
```

TylerLeonhardt commented on src/System.Management.Automation/engine/remoting/common/RemoteSessionNamedPipe.cs:
```diff
@@ -92,11 +99,24 @@ internal static class NamedPipeUtils
+            pipeNameBuilder.Append('.');
+            pipeNameBuilder.Append(CleanAppDomainNameForPipeName(appDomainName));
+            pipeNameBuilder.Append('.');
+            pipeNameBuilder.Append(proc.ProcessName);
```
a styling thing... not sure what's consistent:

```csharp
pipeNameBuilder.Append('.')
    .Append(proc.Id.ToString(CultureInfo.InvariantCulture))
    .Append('.')
    .Append(CleanAppDomainNameForPipeName(appDomainName))
    .Append('.')
    .Append(proc.ProcessName);
```

SteveL-MSFT committed:
[feature]
address additional comments from Paul and Tyler

daxian-dbw commented on src/System.Management.Automation/engine/remoting/common/RemoteSessionNamedPipe.cs:
```diff
@@ -24,9 +25,15 @@ internal static class NamedPipeUtils
+        internal const string DefaultAppDomainName = "Default";
         internal const string NamedPipeNamePrefix = "PSHost.";
+#if UNIX
+        internal const string NamedPipeNamePrefixSearch = "CoreFxPipe_PSHost*";
```
I guess dotnet adds the `CoreFxPipe_` prefix, right?

iSazonov replied:
I found it here https://github.com/dotnet/corefx/blob/a10890f4ffe0fadf090c922578ba0e606ebdd16c/src/System.IO.Pipes/src/System/IO/Pipes/PipeStream.Unix.cs#L33

TylerLeonhardt replied:
Yes. We have the same logic in PowerShell Editor Services to handle the prefix. 

Funny how there's a length limit to the name of the pipe and CoreFX adds this silly prefix.

daxian-dbw commented on src/System.Management.Automation/engine/remoting/common/RemoteSessionNamedPipe.cs:
```diff
@@ -92,11 +99,32 @@ internal static class NamedPipeUtils
+            if (charsToTrim > 0)
+            {
+                pipeNameBuilder.Remove(MaxNamedPipeNameSize + 1, charsToTrim);
+            }
```
This operation might cause the pipe file to not be found by the pipe discovery code above. I guess that would be rare and we can deal with it such an issue comes up?

PaulHigin replied:
I agree.  We should ensure the pipe file pathname is parse-able or it is an error (to log?).

SteveL-MSFT replied:
@daxian-dbw I considered that, but it does seem rare unless we start having hosts with long filenames like `powershell_ise`...

Due to the length restriction on the pipe path and the current design to embed metadata in the filename, I don't see a clear solution to this other than creating a different file that maps guid named pipe files.  But I would defer that until there's an actual issue.

daxian-dbw replied:
I'm fine that we leave it as is until a real issue comes up, which should be rare.

PaulHigin commented on src/System.Management.Automation/engine/remoting/commands/EnterPSHostProcessCommand.cs:
```diff
@@ -565,14 +574,45 @@ internal static IReadOnlyCollection<PSHostProcessInfo> GetAppDomainNamesFromProc
-                                        new PSHostProcessInfo(pName, id, appDomainName));
+                                    Process process = null;
+
+                                    // Only return instances that aren't ourselves since you cannot enter yourself
```
We should always return all processes that host PowerShell.  This cmdlet lists all processes that host PowerShell, not just processes that can be entered.  It is useful to get the complete list.

SteveL-MSFT replied:
While debugging, it was annoying to always check the current pid to know which I can actually enter.  Is there a real use case here where returning the process that you're current running `Get-PSHostProcessInfo` is useful since you can't do anything with it?

SteveL-MSFT replied:
Ok, it's fair for enumerating all PSHosts without intent to enter them all.  I'll defer this and just revert the filtering code.

PaulHigin commented on src/System.Management.Automation/engine/remoting/commands/EnterPSHostProcessCommand.cs:
```diff
@@ -565,14 +574,45 @@ internal static IReadOnlyCollection<PSHostProcessInfo> GetAppDomainNamesFromProc
+                                                // best effort to cleanup
+                                            }
+                                        }
+                                        else if (process.ProcessName.Equals(pName, StringComparison.Ordinal))
```
I don't think this name check is necessary.  If the named pipe file is old or bogus then the process null check above should catch it.

daxian-dbw replied:
I think the intention is to cover the case that a process ID gets reused by another process.

SteveL-MSFT replied:
I'm thinking of the case where a pid gets reused by a different process

SteveL-MSFT replied:
But starttime currently is only decorative for uniqueness and not for validating if it's the same pwsh or a new one reusing the pid

PaulHigin commented on src/System.Management.Automation/engine/remoting/common/RemoteSessionNamedPipe.cs:
```diff
@@ -92,11 +99,32 @@ internal static class NamedPipeUtils
+                // on non-Windows systems, so we'll convert the starttime to hex and just take the first 8 characters.
+                .Append(proc.StartTime.ToFileTime().ToString("X8").Substring(1,8))
+#else
+                .Append(proc.StartTime.ToFileTime().ToString(CultureInfo.InvariantCulture))
```
We want the name to be as unique as possible.  If there is no length limitation on Windows then we should use the full start time.

SteveL-MSFT replied:
As far as I know, the starttime segment of the filename is just for uniqueness and not processed.  We could have the same code for both rather than having different code for Windows vs Unix.  8 hex characters seems sufficient.  What do you think @PaulHigin ?

PaulHigin replied:
As long as it is unique enough.  On Linux we limit connections to processes created by current user.  For Windows it is system wide.  I think we want to err on the side caution for Windows.

SteveL-MSFT replied:
8 hex characters gives 32-bits of uniqueness which seems sufficient.  FILETIME is int64, but that seems overkill particularly when most of the digits are likely zeros.

PaulHigin commented on src/System.Management.Automation/engine/remoting/commands/EnterPSHostProcessCommand.cs:
```diff
@@ -565,14 +574,45 @@ internal static IReadOnlyCollection<PSHostProcessInfo> GetAppDomainNamesFromProc
                                     }
                                 }
```
This would indicate a malformed named pipe file.  I think it would still be useful to show it, it may indicate malicious use.

daxian-dbw replied:
With the current implementation, we won't be showing it because `id = -1` in that case, and it will cause an exception when calling `Process.GetProcessById(id)` in the `if (id != currentPid)` block.

SteveL-MSFT replied:
We could emit a verbose message and continue

PaulHigin commented on src/System.Management.Automation/engine/remoting/commands/EnterPSHostProcessCommand.cs:
```diff
@@ -393,7 +394,12 @@ public sealed class GetPSHostProcessInfoCommand : PSCmdlet
         private const string ProcessNameParameterSet = "ProcessNameParameterSet";
+
+#if UNIX
+        private static readonly string NamedPipePath = Path.GetTempPath();
```
The module analysis cache folder does sound like a better location.  We just want it to be User specific.

SteveL-MSFT replied:
I'll add a comment here that the path is dictated by corefx, this is here only for search purposes

daxian-dbw commented:
As for the domain name, do we really need it for the pipe name? Is multiple domain is going to be supported in .NET Core 3.0?

I have no problem keeping it for now until we get an accurate answer for the app domain support.

SteveL-MSFT commented on src/System.Management.Automation/engine/remoting/commands/EnterPSHostProcessCommand.cs:
```diff
@@ -524,24 +530,27 @@ internal static IReadOnlyCollection<PSHostProcessInfo> GetAppDomainNamesFromProc
+            {
+                namedPipes.Add(Path.Combine(pipeFileInfo.DirectoryName, pipeFileInfo.Name));
+            }
+            int currentPid = System.Diagnostics.Process.GetCurrentProcess().Id;
```
Compiler complains as there is a `Process` member to this cmdlet

SteveL-MSFT replied:
line 704 doesn't need it because the `#if` prevents it from being compiled

SteveL-MSFT commented on src/System.Management.Automation/engine/remoting/commands/EnterPSHostProcessCommand.cs:
```diff
@@ -565,14 +574,45 @@ internal static IReadOnlyCollection<PSHostProcessInfo> GetAppDomainNamesFromProc
+                                        }
+
+                                        if (process == null)
+                                        {
```
@daxian-dbw the `process == null` check can't be in the `try` block as `GetProcessById()` throws if the process doesn't exist.

daxian-dbw replied:
Good point :)

SteveL-MSFT commented:
@daxian-dbw best I can tell, AppDomains are deprecated in .Net Core and not in the roadmap I can see for 3.0.  So it may be fine to remove it for non-Windows.  We'll have to keep it for Windows so that Windows PowerShell and PowerShell Core interop still works however.

SteveL-MSFT commented on src/System.Management.Automation/engine/remoting/commands/EnterPSHostProcessCommand.cs:
```diff
@@ -565,14 +574,45 @@ internal static IReadOnlyCollection<PSHostProcessInfo> GetAppDomainNamesFromProc
+                                    {
+                                        try
+                                        {
+                                            process = System.Diagnostics.Process.GetProcessById(id);
```
Compiler can't differentiate between Process class and Process parameter

SteveL-MSFT commented:
To retain interoperability between Windows PowerShell and keep the code similar between Windows and non-Windows, I'll shorten the DefaultAppDomainName to 'none' but will keep it in the pipe name.

SteveL-MSFT committed:
[feature]
address Paul and Dongbo's feedback

daxian-dbw committed:
Remove extra new lines

daxian-dbw reviewed (approved):
> We'll have to keep it for Windows so that Windows PowerShell and PowerShell Core interop still works however.

Then we should keep the AppDomain field on all platforms. `none` is short enough and it's better to keep the same schema.

PaulHigin commented:
@SteveL-MSFT I forgot about maintaining compatibility with Windows PowerShell.  We definitely need to ensure we can still list/parse/connect to those hosts.

SteveL-MSFT commented:
Ugh, with the changes I've made since, I can't connect to/from Windows PowerShell and PowerShell Core.  Debugging.

Commit:
[feature]
revert defaultappdomainname and starttime format on Windows as it's needed to interop with Windows PowerShell

SteveL-MSFT commented:
Looks like DefaultAppDomain is necessary on Windows as there is a check for that string and the starttime needs to stay as filetime integers on Windows to support interop between Windows PowerShell and PowerShell Core.  I believe everything else is addressed.  @PaulHigin  can you take another look?

PaulHigin commented:
@SteveL-MSFT  Yeah, for Windows we need to keep the DomainName to be able to connect to WindowsPowerShell host, and that means keeping the default domain name for Windows platform pipe names.

SteveL-MSFT commented:
@PaulHigin yes, the recent commit has the same DefaultAppDomainName and starttime format on Windows and I verified manually it works from and to Windows PowerShell

PaulHigin reviewed (approved):
LGTM

I wonder if we should add a test to ensure WindowsPowerShell is listed by Get-PSHostProcessInfo?  WindowsPowerShell is always installed on Windows so it shouldn't be too hard to run powershell.exe and check.

SteveL-MSFT commented:
@PaulHigin good suggestion.  Let me add a test for Windows PowerShell interop.

Commit:
[feature]
added tests to validate interop with Windows PowerShell

daxian-dbw commented:
@SteveL-MSFT, there are tests about `Enter-PSHostProcess` failing in AppVeyor, can you please take a look?

SteveL-MSFT commented:
@daxian-dbw it appears to be a problem with redirected input with PSReadLine and only on AppVeyor.  We're moving off AppVeyor (eventually), so I think we can ignore this otherwise we'll have to special case this by perhaps removing PSReadLine for this test

daxian-dbw commented:
Given @SteveL-MSFT's clarification on the AppVeyor test failure, I'm merging this PR.
